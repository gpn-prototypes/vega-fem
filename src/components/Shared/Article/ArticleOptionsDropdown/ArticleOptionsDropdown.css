.ArticleOptionsDropdown {
  &__menu {
    z-index: var(--layout-z-portal);

    padding: var(--space-xs) 0;

    display: flex;
    flex-direction: column;

    box-sizing: border-box;
    min-width: 150px;

    background: var(--color-bg-default);

    border: var(--control-border-width) solid var(--color-bg-border);
    border-radius: var(--control-radius);
    box-shadow: 0 4px 16px rgba(0, 32, 51, 0.18), 0 2px 2px rgba(0, 0, 0, 0.04);
  }

  &__menuItem {
    text-align: left;
  }
}
