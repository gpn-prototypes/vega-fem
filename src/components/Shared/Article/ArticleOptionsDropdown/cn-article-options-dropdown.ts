import bemCn from 'bem-cn';

export const cnArticleOptionsDropdown = bemCn('ArticleOptionsDropdown');
