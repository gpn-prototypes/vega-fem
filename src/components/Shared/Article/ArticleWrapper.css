.ArticleWrapper__text-field {
  grid-column: 1 / 4;
}

.ArticleWrapper__row {
  margin-top: var(--space-2xs);
  grid-column-gap: var(--space-2xs);
}
