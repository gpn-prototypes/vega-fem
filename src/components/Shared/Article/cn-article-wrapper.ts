import bemCn from 'bem-cn';

export const cnArticleWrapper = bemCn('ArticleWrapper');
