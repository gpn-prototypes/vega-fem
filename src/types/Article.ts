export default interface Article {
  id?: string | number;
  name?: string;
  caption?: string;
  valueTotal?: number;
  value?: ArticleValues[] | number;
  unit?: string;
}

export interface ArticleValues {
  year?: number;
  value: number;
}
