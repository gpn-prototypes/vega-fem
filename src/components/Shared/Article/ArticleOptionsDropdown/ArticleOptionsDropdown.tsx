import React from 'react';
import { Button, Dropdown, IconKebab, usePortal } from '@gpn-prototypes/vega-ui';

import Article from '../../../../../types/Article';

import { Menu } from './Menu/Menu';
import { cnArticleOptionsDropdown } from './cn-article-options-dropdown';

import './ArticleOptionsDropdown.css';

export interface ArticleOptions {
  article: Article;
  updateArticle: (article: Article) => void;
  deleteArticle: (article: Article) => void;
}

export const ArticleOptionsDropdown = ({
  article,
  updateArticle,
  deleteArticle,
}: ArticleOptions) => {
  // const buttonRef = useRef(null);

  const [isOpen, setIsOpen] = React.useState(false);
  const { portal } = usePortal();

  return (
    <>
      <Dropdown
        placement="bottom-start"
        portal={portal}
        isOpen={isOpen}
        onToggle={(nextState): void => {
          setIsOpen(nextState);
        }}
      >
        <Dropdown.Trigger>
          {({ toggle, props: { ref, ...triggerProps } }): React.ReactNode => (
            <Button
              type="button"
              size="s"
              label="Опции"
              onlyIcon
              iconLeft={IconKebab}
              view="ghost"
              ref={ref}
              onClick={toggle}
              {...triggerProps}
            />
          )}
        </Dropdown.Trigger>
        <Dropdown.Menu>
          {({ props: menuProps }): React.ReactNode => (
            <div className={cnArticleOptionsDropdown('menu')} {...menuProps}>
              <Menu
                onClose={() => setIsOpen(false)}
                article={article}
                deleteArticle={deleteArticle}
                updateArticle={updateArticle}
              />
            </div>
          )}
        </Dropdown.Menu>
      </Dropdown>
    </>
  );
};
