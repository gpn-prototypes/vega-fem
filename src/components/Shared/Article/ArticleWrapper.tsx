import React, { useCallback, useState } from 'react';
import { Form, TextField } from '@gpn-prototypes/vega-ui';

import Article, { ArticleValues } from '../../../../types/Article';
import { prepareStringForBack, spreadValue } from '../../../helpers/spreadValue';
import { cnVegaFormCustom } from '../../../styles/VegaFormCustom/cn-vega-form-custom';
import { cnGroupWrapper } from '../../Macroparameters/MacroparameterSetWrapper/GroupWrapper/cn-group-wrapper';

import { ArticleOptionsDropdown } from './ArticleOptionsDropdown/ArticleOptionsDropdown';
import { cnArticleWrapper } from './cn-article-wrapper';

import '../../../styles/BlockWrapper/BlockWrapper.css';
import '../../Macroparameters/MacroparameterSetWrapper/GroupWrapper/GroupWrapper.css';
import './ArticleWrapper.css';

export interface ArticleWrapperProps {
  article: Article;
  fullWidth?: boolean;
  updateArticleValueCallback?: (updatedArticle: Article) => void;
  updateArticleCallback: (updateArticle: Article) => void;
  deleteArticleCallback: (deleteArticle: Article) => void;
  onFocusCallback?: (article: Article) => void;
  highlightArticleClear?: () => void;
}

export const ArticleWrapper = ({
  article,
  updateArticleValueCallback,
  updateArticleCallback,
  deleteArticleCallback,
  fullWidth,
  onFocusCallback,
  highlightArticleClear,
}: ArticleWrapperProps) => {
  const [values, setValues] = useState(article?.value as ArticleValues[]);
  const [spreaded, setSpreaded] = useState<boolean>(true);

  const editValues = (e: any): void => {
    if (spreaded) setSpreaded(false);
    setValues([{ value: e.e.target.value }]);
  };

  const loseFocus = (e: any) => {
    // TODO: change any
    if (e.key === 'Enter') {
      (e.target as HTMLElement).blur();
    }
  };

  const blurHandle = useCallback(() => {
    if (updateArticleValueCallback) {
      updateArticleValueCallback({
        ...article,
        ...{
          value:
            typeof values[0]?.value === 'string'
              ? prepareStringForBack(values[0]?.value)
              : values[0]?.value,
        },
      });
      setSpreaded(true);
    }
    if (highlightArticleClear) {
      highlightArticleClear();
    }
  }, [values, updateArticleValueCallback, article, highlightArticleClear]);

  const onFocusHandler = useCallback(() => {
    if (onFocusCallback) {
      onFocusCallback(article);
    }
  }, [onFocusCallback, article]);

  const displayValue = (value: string | number | null | undefined): string => {
    if (value) {
      if (spreaded) return spreadValue(value);
      return value?.toString();
    }
    return '';
  };

  return (
    <Form.Row
      className={cnVegaFormCustom('form-row', { width: fullWidth && 'full-width' })}
      space="m"
    >
      <Form.Field className={cnGroupWrapper('body-content')}>
        <Form.Label space="2xs" data-testid="label">
          {article?.caption}
        </Form.Label>
        <Form.Row col="4" className={cnArticleWrapper('row')}>
          <Form.Field className={cnArticleWrapper('text-field')}>
            <TextField
              size="s"
              width="full"
              id={`article_${article?.name}_${article?.id}`}
              placeholder="Значение"
              rightSide={article?.unit}
              value={displayValue(values[0]?.value)}
              onBlur={blurHandle}
              onChange={(e: any) => editValues(e)}
              onKeyDown={(e) => loseFocus(e)}
              onFocus={onFocusHandler}
              data-testid="input"
            />
          </Form.Field>
          <Form.Field>
            <ArticleOptionsDropdown
              article={article}
              updateArticle={updateArticleCallback}
              deleteArticle={deleteArticleCallback}
            />
          </Form.Field>
        </Form.Row>
      </Form.Field>
    </Form.Row>
  );
};
