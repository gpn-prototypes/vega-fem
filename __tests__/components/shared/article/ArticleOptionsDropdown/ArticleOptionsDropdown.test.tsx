import * as React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';

import { ArticleOptionsDropdown } from '@/components/Shared/Article/ArticleOptionsDropdown/ArticleOptionsDropdown';
import Article from '@/types/Article';

let fakeArticle: Article;
beforeAll(() => {
  fakeArticle = {
    id: 1,
    name: 'fake_article',
    caption: 'fake article',
    value: [{ year: 2015, value: 100 }],
    unit: 'руб.',
  };
});

describe('ArticleOptionsDropdown', () => {
  test('рендерится корректно', () => {
    const tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = renderer
      .create(
        <ArticleOptionsDropdown
          updateArticle={jest.fn()}
          deleteArticle={jest.fn()}
          article={fakeArticle}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
