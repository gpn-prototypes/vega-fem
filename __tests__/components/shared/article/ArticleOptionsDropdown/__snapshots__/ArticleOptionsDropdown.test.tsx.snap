// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<button
  className="Button Button_size_s Button_view_ghost Button_form_default Button_withIcon Button_onlyIcon MixFocus"
  onClick={[Function]}
  title="Опции"
  type="button"
>
  <span
    className="Icon Icon_size_xs IconKebab Button-Icon"
  >
    <svg
      className="Icon-Svg"
      viewBox="0 0 16 16"
    >
      <path
        d="M8 4.5c.83 0 1.5-.67 1.5-1.5S8.83 1.5 8 1.5 6.5 2.17 6.5 3 7.17 4.5 8 4.5zm0 2c-.83 0-1.5.67-1.5 1.5S7.17 9.5 8 9.5 9.5 8.83 9.5 8 8.83 6.5 8 6.5zm0 5c-.83 0-1.5.67-1.5 1.5s.67 1.5 1.5 1.5 1.5-.67 1.5-1.5-.67-1.5-1.5-1.5z"
      />
    </svg>
  </span>
</button>
`;

exports[`ArticleOptionsDropdown рендерится корректно 1`] = `
<button
  className="Button Button_size_s Button_view_ghost Button_form_default Button_withIcon Button_onlyIcon MixFocus"
  onClick={[Function]}
  title="Опции"
  type="button"
>
  <span
    className="Icon Icon_size_xs IconKebab Button-Icon"
  >
    <svg
      className="Icon-Svg"
      viewBox="0 0 16 16"
    >
      <path
        d="M8 4.5c.83 0 1.5-.67 1.5-1.5S8.83 1.5 8 1.5 6.5 2.17 6.5 3 7.17 4.5 8 4.5zm0 2c-.83 0-1.5.67-1.5 1.5S7.17 9.5 8 9.5 9.5 8.83 9.5 8 8.83 6.5 8 6.5zm0 5c-.83 0-1.5.67-1.5 1.5s.67 1.5 1.5 1.5 1.5-.67 1.5-1.5-.67-1.5-1.5-1.5z"
      />
    </svg>
  </span>
</button>
`;
